Problem Statement - Identify key variables that can be used by the consumer finance company can use to identify applicants that are likely to default on a loan.


In [1]:
# Loading the data given
import pandas as pd
cust_details = pd.read_csv('/Users/zsp/Downloads/loan.csv',low_memory=False)
cust_details.info()
cust_details.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Columns: 111 entries, id to total_il_high_credit_limit
dtypes: float64(74), int64(13), object(24)
memory usage: 33.6+ MB


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   

  int_rate  installment grade sub_grade  ... num_tl_90g_dpd_24m  \
0   10.65%       162.87     B        B2  ...                NaN   
1   15.27%        59.83     C        C4  ...                NaN   
2   15.96%        84.33     C        C5  ...                NaN   

  num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75 pub_rec_bankruptcies  \
0                NaN            NaN               NaN                  0.0   
1                NaN            NaN               NaN                  0.0   
2                NaN            NaN               NaN                  0.0   

  tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0       0.0             NaN               NaN            NaN   
1       0.0             NaN               NaN            NaN   
2       0.0             NaN               NaN            NaN   

  total_il_high_credit_limit  
0                        NaN  
1                        NaN  
2                        NaN  

[3 rows x 111 columns]

The inital dataset contains 111 columns and a total of 39717 row entries. First we identify if all those 111 columns contain usefull information - if they do not then we remove them from further consideration.

In [2]:
print(cust_details.isnull().sum())

id                                0
member_id                         0
loan_amnt                         0
funded_amnt                       0
funded_amnt_inv                   0
                              ...  
tax_liens                        39
tot_hi_cred_lim               39717
total_bal_ex_mort             39717
total_bc_limit                39717
total_il_high_credit_limit    39717
Length: 111, dtype: int64


In [3]:
# remove columns where no data is provided
cust_details.dropna(how='all', axis=1, inplace=True)

In [4]:
cust_details.info()
cust_details.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          39717 non-null  int64  
 1   member_id                   39717 non-null  int64  
 2   loan_amnt                   39717 non-null  int64  
 3   funded_amnt                 39717 non-null  int64  
 4   funded_amnt_inv             39717 non-null  float64
 5   term                        39717 non-null  object 
 6   int_rate                    39717 non-null  object 
 7   installment                 39717 non-null  float64
 8   grade                       39717 non-null  object 
 9   sub_grade                   39717 non-null  object 
 10  emp_title                   37258 non-null  object 
 11  emp_length                  38642 non-null  object 
 12  home_ownership              39717 non-null  object 
 13  annual_inc                  397

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   

  int_rate  installment grade sub_grade  ... next_pymnt_d last_credit_pull_d  \
0   10.65%       162.87     B        B2  ...          NaN             May-16   
1   15.27%        59.83     C        C4  ...          NaN             Sep-13   
2   15.96%        84.33     C        C5  ...          NaN             May-16   

  collections_12_mths_ex_med  policy_code application_type acc_now_delinq  \
0                        0.0            1       INDIVIDUAL              0   
1                        0.0            1       INDIVIDUAL              0   
2                        0.0            1       INDIVIDUAL              0   

  chargeoff_within_12_mths delinq_amnt pub_rec_bankruptcies tax_liens  
0                      0.0           0                  0.0       0.0  
1                      0.0           0                  0.0       0.0  
2                      0.0           0                  0.0       0.0  

[3 rows x 57 columns]

In [5]:
# Remove columns with constant values as these will not provide usefull information for the analysis
# # cust_details.to_csv('loan_modified.csv')
# print(cust_details.columns[cust_details.nunique() <= 1])
# # print(cust_details.loc[:, list(cust_details.columns[cust_details.nunique() <= 1])])
cust_details_mod=cust_details.drop(cust_details.loc[:, list(cust_details.columns[cust_details.nunique() <= 1])], axis=1)

In [6]:
cust_details_mod.info()
cust_details_mod.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 48 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       39717 non-null  int64  
 1   member_id                39717 non-null  int64  
 2   loan_amnt                39717 non-null  int64  
 3   funded_amnt              39717 non-null  int64  
 4   funded_amnt_inv          39717 non-null  float64
 5   term                     39717 non-null  object 
 6   int_rate                 39717 non-null  object 
 7   installment              39717 non-null  float64
 8   grade                    39717 non-null  object 
 9   sub_grade                39717 non-null  object 
 10  emp_title                37258 non-null  object 
 11  emp_length               38642 non-null  object 
 12  home_ownership           39717 non-null  object 
 13  annual_inc               39717 non-null  float64
 14  verification_status   

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   

  int_rate  installment grade sub_grade  ... total_rec_prncp total_rec_int  \
0   10.65%       162.87     B        B2  ...         5000.00        863.16   
1   15.27%        59.83     C        C4  ...          456.46        435.17   
2   15.96%        84.33     C        C5  ...         2400.00        605.67   

  total_rec_late_fee  recoveries collection_recovery_fee last_pymnt_d  \
0                0.0        0.00                    0.00       Jan-15   
1                0.0      117.08                    1.11       Apr-13   
2                0.0        0.00                    0.00       Jun-14   

  last_pymnt_amnt next_pymnt_d last_credit_pull_d pub_rec_bankruptcies  
0          171.62          NaN             May-16                  0.0  
1          119.66          NaN             Sep-13                  0.0  
2          649.91          NaN             May-16                  0.0  

[3 rows x 48 columns]

While examing the rows I found that rows where the 'total_pymt' was zero more that 10 other columns in those rows also contained zeros. Which seemed odd since this data seemed to extracted on 2016 for 2008-2011 data and unless there was a loss of data there should have been some entry in these columns.

In [7]:
#remove rows where the relevant data is missing
# cust_details_mod['total_pymnt_inv'].isnull()
cust_details_mod.loc[cust_details_mod['total_pymnt'] == 0]

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
26     1069800    1304679      15000        15000           8725.0   
2724   1032519    1262161       8000         8000           3975.0   
3985   1009833    1236591      18000        18000          18000.0   
6967    889684    1106242       2200         2200           2200.0   
8707    859150    1071830      35000        35000          34975.0   
14220   768963     954342       3450         3450           3450.0   
14642   762630     963168       5000         5000           4175.0   
16301   734671     931223      10000        10000           8600.0   
16719   727583     923109       5550         5550           5550.0   
18316   699321     890868      10000        10000          10000.0   
27644   550066     709061       2000         2000           2000.0   
34237   454682     563309       6000         6000           5950.0   
34420   451820     557058       4000         4000           3975.0   
35186   439641     531378      20000        20000          19050.0   
36206   417817     487349       7800         7800           7725.0   
36937   392751     429885      25000        25000          11625.0   

             term int_rate  installment grade sub_grade  ... total_rec_prncp  \
26      36 months   14.27%       514.64     C        C2  ...             0.0   
2724    36 months   14.65%       275.96     C        C3  ...             0.0   
3985    60 months   19.03%       467.23     E        E2  ...             0.0   
6967    60 months   19.42%        57.58     E        E3  ...             0.0   
8707    60 months   13.99%       814.21     C        C3  ...             0.0   
14220   36 months   11.99%       114.58     B        B5  ...             0.0   
14642   36 months   16.49%       177.00     D        D3  ...             0.0   
16301   36 months   13.43%       339.02     C        C3  ...             0.0   
16719   60 months   13.06%       126.46     C        C2  ...             0.0   
18316   36 months   13.80%       340.81     C        C4  ...             0.0   
27644   36 months   11.49%        65.95     B        B4  ...             0.0   
34237   36 months   14.61%       206.86     D        D1  ...             0.0   
34420   36 months   12.87%       134.54     C        C1  ...             0.0   
35186   36 months   17.39%       716.95     E        E4  ...             0.0   
36206   36 months   10.95%       255.17     B        B1  ...             0.0   
36937   36 months   13.79%       851.89     C        C5  ...             0.0   

      total_rec_int total_rec_late_fee  recoveries collection_recovery_fee  \
26              0.0                0.0         0.0                     0.0   
2724            0.0                0.0         0.0                     0.0   
3985            0.0                0.0         0.0                     0.0   
6967            0.0                0.0         0.0                     0.0   
8707            0.0                0.0         0.0                     0.0   
14220           0.0                0.0         0.0                     0.0   
14642           0.0                0.0         0.0                     0.0   
16301           0.0                0.0         0.0                     0.0   
16719           0.0                0.0         0.0                     0.0   
18316           0.0                0.0         0.0                     0.0   
27644           0.0                0.0         0.0                     0.0   
34237           0.0                0.0         0.0                     0.0   
34420           0.0                0.0         0.0                     0.0   
35186           0.0                0.0         0.0                     0.0   
36206           0.0                0.0         0.0                     0.0   
36937           0.0                0.0         0.0                     0.0   

      last_pymnt_d last_pymnt_amnt next_pymnt_d last_credit_pull_d  \
26             NaN             0.0          NaN             May

In [8]:
cust_details_mod.drop(cust_details_mod.loc[cust_details_mod['total_pymnt']==0].index, inplace=True)

In [9]:
# cust_details_mod.to_csv('loan_modified_2.csv')

In [10]:
cust_details_mod.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 39701 entries, 0 to 39716
Data columns (total 48 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       39701 non-null  int64  
 1   member_id                39701 non-null  int64  
 2   loan_amnt                39701 non-null  int64  
 3   funded_amnt              39701 non-null  int64  
 4   funded_amnt_inv          39701 non-null  float64
 5   term                     39701 non-null  object 
 6   int_rate                 39701 non-null  object 
 7   installment              39701 non-null  float64
 8   grade                    39701 non-null  object 
 9   sub_grade                39701 non-null  object 
 10  emp_title                37243 non-null  object 
 11  emp_length               38627 non-null  object 
 12  home_ownership           39701 non-null  object 
 13  annual_inc               39701 non-null  float64
 14  verification_status   

On further consideration I decided to remove the rows where the loan_status is 'current' as these are still active it gives us little information on whether or not these customers would default or not.

In [11]:
cust_details_mod.drop(cust_details_mod.loc[cust_details_mod['loan_status']=='Current'].index, inplace=True)

In [12]:
cust_details_mod.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38561 entries, 0 to 39716
Data columns (total 48 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       38561 non-null  int64  
 1   member_id                38561 non-null  int64  
 2   loan_amnt                38561 non-null  int64  
 3   funded_amnt              38561 non-null  int64  
 4   funded_amnt_inv          38561 non-null  float64
 5   term                     38561 non-null  object 
 6   int_rate                 38561 non-null  object 
 7   installment              38561 non-null  float64
 8   grade                    38561 non-null  object 
 9   sub_grade                38561 non-null  object 
 10  emp_title                36176 non-null  object 
 11  emp_length               37529 non-null  object 
 12  home_ownership           38561 non-null  object 
 13  annual_inc               38561 non-null  float64
 14  verification_status   

After the data cleaning steps I was left with the above 48 columns and 38561 rows.
Next I segment the data based on 'loan status'

In [13]:
df_pv=pd.pivot_table(data=cust_details_mod,index='loan_status',aggfunc='size')
print((df_pv/df_pv.sum())*100)


loan_status
Charged Off    14.550971
Fully Paid     85.449029
dtype: float64


From the above figure we can see that 85% repaid their loans but the remaining 16% defaulted.
Next, I will go through the list of variabale and try and figure out which ones can be used to predict defaulters.
First we examin the type of home ownership

In [14]:
pd.crosstab(cust_details_mod.loan_status,cust_details_mod.home_ownership,margins=True,normalize='index')

home_ownership  MORTGAGE      NONE     OTHER       OWN      RENT
loan_status                                                     
Charged Off     0.414365  0.000000  0.003208  0.078596  0.503832
Fully Paid      0.445948  0.000091  0.002428  0.076844  0.474689
All             0.441353  0.000078  0.002541  0.077099  0.478929

Comparing the percentage of Mortage and renters in both 'Charged off' & 'Fully Paid' categories we can see
that the majority of borowers fall into these 2 categories and therefore cannot be used identify defaulters

In [15]:
pd.crosstab(cust_details_mod.loan_status,cust_details_mod.verification_status,margins=True,normalize='index')

verification_status  Not Verified  Source Verified  Verified
loan_status                                                 
Charged Off              0.380324         0.255035  0.364641
Fully Paid               0.441639         0.250167  0.308194
All                      0.432717         0.250875  0.316408

Verification also does not seem to be an indicator

In [16]:
pd.crosstab(cust_details_mod.purpose,cust_details_mod.loan_status,margins=True,normalize='columns')

loan_status         Charged Off  Fully Paid       All
purpose                                              
car                    0.028515    0.040637  0.038873
credit_card            0.096596    0.136115  0.130365
debt_consolidation     0.491356    0.463976  0.467960
educational            0.009980    0.008164  0.008428
home_improvement       0.061486    0.076722  0.074505
house                  0.010515    0.009347  0.009517
major_purchase         0.039565    0.058513  0.055756
medical                0.018891    0.017451  0.017660
moving                 0.016396    0.014689  0.014937
other                  0.112636    0.098088  0.100205
renewable_energy       0.003386    0.002519  0.002645
small_business         0.084299    0.038816  0.045435
vacation               0.009446    0.009772  0.009725
wedding                0.016931    0.025190  0.023988

From the above data we can see that majority of borrowers are take a loan for debt_consolidation. 
Majority of defaulters fall into this category ~50%. So this means that these applicants should be scrutinized a bit more.

In [18]:
pd.crosstab(cust_details_mod.purpose,cust_details_mod.loan_status,margins=True,normalize='index')

loan_status         Charged Off  Fully Paid
purpose                                    
car                    0.106738    0.893262
credit_card            0.107818    0.892182
debt_consolidation     0.152785    0.847215
educational            0.172308    0.827692
home_improvement       0.120084    0.879916
house                  0.160763    0.839237
major_purchase         0.103256    0.896744
medical                0.155653    0.844347
moving                 0.159722    0.840278
other                  0.163561    0.836439
renewable_energy       0.186275    0.813725
small_business         0.269977    0.730023
vacation               0.141333    0.858667
wedding                0.102703    0.897297
All                    0.145510    0.854490


Does interest rate play a role?

In [19]:
cust_details_mod['int_rate']=cust_details_mod['int_rate'].str.replace(r'\D', '', regex=True).astype(float)/100

In [20]:
import numpy as np
bins_list=np.arange(cust_details_mod['int_rate'].min(), cust_details_mod['int_rate'].max(), 3).tolist()
print(np.size(bins_list))
print((bins_list))

7
[5.42, 8.42, 11.42, 14.42, 17.42, 20.42, 23.42]


In [21]:
cust_details_mod['int_rate_bin'] = pd.cut(cust_details_mod['int_rate'],bins_list)

In [22]:
cust_details_mod['int_rate_bin']

0         (8.42, 11.42]
1        (14.42, 17.42]
2        (14.42, 17.42]
3        (11.42, 14.42]
5          (5.42, 8.42]
              ...      
39712      (5.42, 8.42]
39713     (8.42, 11.42]
39714      (5.42, 8.42]
39715      (5.42, 8.42]
39716    (11.42, 14.42]
Name: int_rate_bin, Length: 38561, dtype: category
Categories (6, interval[float64, right]): [(5.42, 8.42] < (8.42, 11.42] < (11.42, 14.42] < (14.42, 17.42] < (17.42, 20.42] < (20.42, 23.42]]

In [23]:
cust_details_mod.pivot_table(index='loan_status',columns='int_rate_bin',aggfunc='size',fill_value=0)

int_rate_bin  (5.42, 8.42]  (8.42, 11.42]  (11.42, 14.42]  (14.42, 17.42]  \
loan_status                                                                 
Charged Off            425            992            1811            1388   
Fully Paid            7305           8298            9604            5088   

int_rate_bin  (17.42, 20.42]  (20.42, 23.42]  
loan_status                                   
Charged Off              750             216  
Fully Paid              1761             327

In [ ]:
pd.crosstab(cust_details_mod['loan_status'],cust_details_mod['int_rate_bin'],margins=True,normalize='index')

In [ ]:
cust_details_mod.columns

In [ ]:
bins_list=np.arange(cust_details_mod.loan_amnt.min()
                    , cust_details_mod.loan_amnt.max(), 5000).tolist()

print(bins_list)

In [ ]:
cust_details_mod['loan_amnt_bin'] = pd.cut(cust_details_mod.loan_amnt,bins_list)
print(cust_details_mod['loan_amnt_bin'])

In [ ]:
bins_list=np.arange(cust_details_mod.funded_amnt.min()
                    , cust_details_mod.funded_amnt.max(), 5000).tolist()

print(bins_list)

In [ ]:
cust_details_mod['funded_amnt_bin'] = pd.cut(cust_details_mod.funded_amnt,bins_list)
print(cust_details_mod['funded_amnt_bin'])

In [ ]:
cust_details_mod.pivot_table(index='loan_status',columns='funded_amnt_bin',aggfunc='size',fill_value=0)

In [ ]:
pd.crosstab(cust_details_mod['loan_status'],cust_details_mod['funded_amnt_bin'],margins=True,normalize='index')

In [ ]:
cust_details_mod.corr()

In [ ]:
print(cust_details_mod.iloc[:, (np.where((cust_details_mod.dtypes == np.int64) | 
                 (cust_details_mod.dtypes == np.float64)))[0]].columns)

In [ ]:
df=cust_details_mod[['loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'int_rate', 'installment', 'annual_inc', 'dti']].copy()

In [ ]:
df.head(5)

In [ ]:
corr=df.corr()

In [ ]:
corr.style.background_gradient(cmap='coolwarm')

In [ ]:
df=cust_details_mod[['loan_status','loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'int_rate', 'delinq_2yrs',
       'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record',
       'open_acc', 'pub_rec', 'revol_bal', 'total_acc', 'out_prncp',
       'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp',
       'total_rec_int', 'total_rec_late_fee', 'recoveries']].copy()

In [ ]:
corr=df.corr()

In [ ]:
corr.style.background_gradient(cmap='coolwarm')

In [ ]:
df=cust_details_mod[['loan_status','id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'int_rate', 'installment', 'annual_inc', 'dti', 'delinq_2yrs',
       'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record',
       'open_acc', 'pub_rec', 'revol_bal', 'total_acc', 'out_prncp',
       'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp',
       'total_rec_int', 'total_rec_late_fee', 'recoveries',
       'collection_recovery_fee', 'last_pymnt_amnt', 'pub_rec_bankruptcies']].copy()

In [ ]:
df_m=df.loc[df['loan_status'] == 0]